File: Bitmasking_for_Sets_states/Q1. Gray Code/Q1_GrayCode.java
import java.util.*;

class Solution {
    public List<Integer> grayCode(int n) {
        List<Integer> result = new ArrayList<>(); // List to store gray code sequence
        result.add(0);                             // Start sequence with 0

        for (int i = 0; i < n; i++) {             // Loop through each bit position
            int size = result.size();             // Current size of sequence
            for (int j = size - 1; j >= 0; j--) { // Loop backwards through current sequence
                result.add(result.get(j) | (1 << i)); // Add new numbers by flipping the i-th bit
            }
        }

        return result;                             // Return final gray code sequence
    }
}

File: Bitmasking_for_Sets_states/Q1. Gray Code/Q1_GrayCode.py
class Solution(object):
    def grayCode(self, n):
        """
        :type n: int
        :rtype: List[int]
        """
        result = [0]               # Start sequence with 0

        for i in range(n):         # Loop through each bit position
            size = len(result)     # Current size of sequence
            for j in range(size - 1, -1, -1):  # Loop backward through current sequence
                result.append(result[j] | (1 << i))  # Flip the i-th bit and add to sequence

        return result              # Return final gray code sequence
